src/compile/encodings.js

```diff
@@ -4,6 +4,32 @@ var varint = require('varint')
 var svarint = require('signed-varint')
 const { Buffer } = require('buffer')
 
+/**
+ * @template T
+ * @typedef {{bytes?:number} & function(T, Buffer, number):Buffer} Encoder
+ */
+/**
+ * @template T
+ * @typedef {{bytes?:number} & function(Buffer, number):T} Decoder
```

Comment:
can you explain this type? normally i assume intersection is used for previously defined types here we are defining all types inline why use intersection?

what does the final type look like?

Reply:
Mostly it's about TS not recognizing following pattern (for good reasons) all that well:

```ts
function decode(buf, offset) {
  // ...
}
decode.bytes = 0
```

Function does not have `bytes` property & that is what TS will tell you. Type declaration at hand uses intersection type to communicate with TS that our function is special as it has `bytes:number` property.

I'd typically avoid such pattern, or otherwise type it as follows in TS syntax:

```ts
interface Decoder<T> {
   (buffer:Buffer, offset:number):T
   bytes?: number
}
```

Which is functionally equivalent to:

```ts
type Decoder<T> = ((buffer:Buffer, offset:number) => T & { bytes?: number })
```

We could obviously have predefined that function extension type and did intersection with it instead:

```ts
type WithBytes = { bytes?: number }
type Decoder<T> =  (buffer:Buffer, offset:number) => T
type DecoderWithByets<T> = Decoder<T> & WithBytes
```

But naming things is hard and sometimes inlining saves you from it 😊

Reply:
can't we avoid that pattern ? its better to fix the code than to build complex types

Reply:
Can we please solve this problem in a different PR?  I think this is way beyond the scope of supporting Uint8Arrays in this module.

Reply:
To be clear, I meant adding the TS comments and then trying to solve the problems they uncover is outside the scope of supporting Uint8Arrays in this module.

Reply:
@achingbrain just so that I understand, should I remove the added comments in this pull or leave them in ?

Reply:
> can't we avoid that pattern ? its better to fix the code than to build complex types

@hugomrdias In theory yes. Probably good idea for new code. However I'm not it's a practical thing to do with existing code base. And absolutely out of scope for this.

Reply:
Please remove them, they can be added in a separate PR, if they are required.

Reply:
Done